src/internal/connector/onedrive/restore.go

```diff
@@ -669,14 +714,70 @@ func createRestoreFolders(
 	return parentFolderID, nil
 }
 
+func copyBufferWithStallCheck(dst io.Writer, src io.Reader, stallTimeout time.Duration) (int64, error) {
+	buffer := make([]byte, copyBufferSize)
+
+	timer := time.NewTimer(stallTimeout)
+	defer timer.Stop()
+
+	type ce struct {
+		complete bool
+		err      error
+	}
+
+	totalCopied := int64(0)
+	cont := make(chan ce)
+
+	for {
+		go func() {
+			n, rerr := src.Read(buffer)
+			if rerr != nil && rerr != io.EOF {
+				cont <- ce{err: clues.Wrap(rerr, "reading data")}
+			}
+
+			_, werr := dst.Write(buffer[:n])
+			if werr != nil && werr != io.EOF {
+				cont <- ce{err: clues.Wrap(werr, "writing data")}
+			}
+
+			totalCopied += int64(n)
+
+			if rerr == io.EOF {
+				// Copy completed successfully
+				cont <- ce{complete: true}
+			}
+
+			cont <- ce{complete: false}
+		}()
+
+		timer.Reset(stallTimeout)
+
+		select {
+		case <-timer.C:
+			return 0, clues.New("copy stalled")
+		case in := <-cont:
+			if in.err != nil {
+				return 0, in.err
```

Comment:
I think this needs a bit more definition on the error handling. For example, if there's a read error it still tries to upload the data and totalCopied is still updated. However, this will return prior to the upload completing and it also returns 0 as the amount of data copied

We haven't really discussed how to handle partial file uploads, but we should at least make sure we don't exit this function with the goroutine from above still running

Reply:
Good catch, I made some last minute changes and missed a few things. 😓

Reply:
I'm dropping the function to copy with stall check. It was just me panicking from the result of https://github.com/alcionai/corso/pull/3492#discussion_r1205775330 .